pkg/api/scylla/v1alpha1/types_nodeconfig.go

```diff
@@ -22,37 +22,6 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-type NodeConfigConditionType string
-
-const (
-	// Reconciled indicates that the NodeConfig is fully deployed and available.
-	NodeConfigReconciledConditionType NodeConfigConditionType = "Reconciled"
-)
-
-type NodeConfigCondition struct {
```

Comment:
the kubebuilder validation for metav1.Condition differs - I am afraid we can't do that in the same API version

Reply:
does the rule about not tightening the validation also apply for statuses? Because this only changes the status.

Reply:
I thinks so - on a storage migration you need to update your object. New status is dropped and the old one stays - then the object is validated - which I assume fails. You can try it out though.

Reply:
It may fail currently because the constructed aggregated node conditions can fail validation (although this also applies to the regular aggregated conditions, see: https://github.com/scylladb/scylla-operator/issues/2102). Imo we could move to Conditions, but not in one release, as the NodeCondition created now wouldn't pass Condition's validation. Imo we could merge this (without NodeCondition->Condition switch) and fix Condition and NodeCondition aggregation to make sure they pass validation and then make a switch in following releases.

Reply:
You can't be sure the sync loop for every single object in a previous release got to store the new status successfully. I'd rather do validation changes  with v1alpha2.

This isn't also required for condition aggregations - you can internally convert the conditions at the start of the sync loop and back at the end
```golang
// TODO: Move NodeConfigCondition to metav1.Condition in the next API version.

type NodeConfigCondition struct {
	// type is the type of the NodeConfig condition.
	Type NodeConfigConditionType `json:"type"`
	// status represents the state of the condition, one of True, False, or Unknown.
	Status corev1.ConditionStatus `json:"status"`
	// observedGeneration represents the .metadata.generation that the condition was set based upon.
	// For instance, if .metadata.generation is currently 12, but the .status.conditions[x].observedGeneration is 9, the condition is out of date
	// with respect to the current state of the instance.
	// +optional
	// +kubebuilder:validation:Minimum=0
	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
	// lastTransitionTime is last time the condition transitioned from one status to another.
	LastTransitionTime metav1.Time `json:"lastTransitionTime"`
	// reason is the reason for condition's last transition.
	Reason string `json:"reason"`
	// message is a human-readable message indicating details about the transition.
	Message string `json:"message"`
}

func (c *NodeConfigCondition) ToMetaV1Condition() metav1.Condition {
	return metav1.Condition{
		Type:               string(c.Type),
		Status:             metav1.ConditionStatus(c.Status),
		ObservedGeneration: c.ObservedGeneration,
		LastTransitionTime: c.LastTransitionTime,
		Reason:             c.Reason,
		Message:            c.Message,
	}
}

func NewNodeConfigCondition(c metav1.Condition) NodeConfigCondition {
	return NodeConfigCondition{
		Type:               NodeConfigConditionType(c.Type),
		Status:             corev1.ConditionStatus(c.Status),
		ObservedGeneration: c.ObservedGeneration,
		LastTransitionTime: c.LastTransitionTime,
		Reason:             c.Reason,
		Message:            c.Message,
	}
}

type NodeConfigConditions []NodeConfigCondition

func (c NodeConfigConditions) ToMetaV1Conditions() []metav1.Condition {
	return slices.ConvertSlice(c, func(from NodeConfigCondition) metav1.Condition {
		return from.ToMetaV1Condition()
	})
}

func NewNodeConfigConditions(cs []metav1.Condition) NodeConfigConditions {
	return slices.ConvertSlice(cs, func(from metav1.Condition) NodeConfigCondition {
		return NewNodeConfigCondition(from)
	})
}
```
```golang
func (ncc *Controller) sync(ctx context.Context, key string) error {
// ...
    statusConditions := status.Conditions.ToMetaV1Conditions()
    if nc.DeletionTimestamp != nil {
        return ncc.updateStatus(ctx, nc, status, statusConditions)
    }
// ...
    err = ncc.updateStatus(ctx, nc, status, statusConditions)
// ...
}
```

Reply:
> This isn't also required for condition aggregations - you can internally convert the conditions at the start of the sync loop and back at the end

Agreed, I wanted to do this get rid of NodeCondition helpers without changing the schema

Reply:
done